lib/scheduler/thread_pool.rb

```diff
@@ -0,0 +1,138 @@
+# frozen_string_literal: true
+
+module Scheduler
+  # ThreadPool manages a pool of worker threads that process tasks from a queue.
+  # It maintains a minimum number of threads and can scale up to a maximum number
+  # when there's more work to be done.
+  class ThreadPool
+    class ShutdownError < StandardError
+    end
+
+    def initialize(min_threads:, max_threads:, idle_time:)
+      raise ArgumentError, "min_threads must be positive" if min_threads <= 0
+      raise ArgumentError, "max_threads must be >= min_threads" if max_threads < min_threads
+      raise ArgumentError, "idle_time must be positive" if idle_time <= 0
+
+      @min_threads = min_threads
+      @max_threads = max_threads
+      @idle_time = idle_time
+
+      @threads = []
+      @queue = Queue.new
+      @mutex = Mutex.new
+      @new_work = ConditionVariable.new
+      @shutdown = false
+
+      # Initialize minimum number of threads
+      @min_threads.times { spawn_thread }
+    end
+
+    def post(&block)
+      raise ShutdownError, "Cannot post work to a shutdown ThreadPool" if shutdown?
+
+      db = RailsMultisite::ConnectionManagement.current_db
+      wrapped_block = wrap_block(block, db)
+
+      @mutex.synchronize do
+        @queue << wrapped_block
+        @new_work.signal
+        spawn_thread if @threads.size < @max_threads
```

Comment:
Should we check if there is an idle thread that can do the work instead of unconditionally spawning a thread whenever work is posted?

Reply:
this is a good point, I worry about cost and stuff going out of sync, having a think about this. I don't really want to shuffle threads between 2 arrays, but I guess we can try this.